core/src/main/java/io/hyperfoil/core/impl/SimulationRunner.java

```diff
@@ -411,6 +416,119 @@ public void addGlobalData(Map<String, GlobalData.Element> globalData) {
       }
    }
 
+   /**
+    * It forces the garbage collection process.
+    * If the JVM arg {@link Properties#GC_CHECK} is set, it will also perform an additional check to ensure that the
+    * GC actually occurred.
+    *
+    * <p>Note: This method uses {@link ManagementFactory#getGarbageCollectorMXBeans()} to gather GC bean information
+    * and checks if at least one GC cycle has completed. It waits for the GC to stabilize within a maximum wait time.
+    * If GC beans are not available, it waits pessimistically for a defined period after invoking {@code System.gc()}.</p>
+    *
+    * @see #runSystemGC()
+    * @see ManagementFactory#getGarbageCollectorMXBeans()
+    */
+   private void runGC() {
+      long start = System.currentTimeMillis();
+
+      if (Properties.getBoolean(Properties.GC_CHECK)) {
+         log.info("Running additional GC check!");
+         // Collect GC beans
+         List<GarbageCollectorMXBean> enabledBeans = new ArrayList<>();
+
+         long beforeGcCount = 0;
+         for (GarbageCollectorMXBean bean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            long count = bean.getCollectionCount();
+            if (count != -1) {
+               enabledBeans.add(bean);
+               beforeGcCount += bean.getCollectionCount();
+            }
+         }
+
+         // Run the GC
+         this.runSystemGC();
+
+         // Make sure the GC actually happened
+         //   a) That at least two collections happened, indicating GC work.
+         //   b) That counter updates have not happened for a while, indicating GC work had ceased.
+
+         final int MAX_WAIT_MS = 10 * 1000;
+
+         if (enabledBeans.isEmpty()) {
+            log.warn("MXBeans can not report GC info. System.gc() invoked, pessimistically waiting {} ms", MAX_WAIT_MS);
+            try {
+               TimeUnit.MILLISECONDS.sleep(MAX_WAIT_MS);
```

Comment:
We could failback to the original behaviour here, instead of letting the user to wait for so long

Reply:
I agree, fixed :heavy_check_mark: 

I also switched the if-else as you were suggesting offline such that the default is the first one